core/trino-main/src/main/java/io/trino/operator/MultiChannelGroupByHash.java

```diff
@@ -846,4 +975,61 @@ public GroupByIdBlock getResult()
                             page.getPositionCount()));
         }
     }
+
+    private abstract class LowCardinalityDictionaryWork<T>
+            implements Work<T>
+    {
+        protected final Page page;
+        protected final int maxCardinality;
+        protected final int[] dictionarySizes;
+        protected final DictionaryBlock[] blocks;
+        protected int lastPosition;
+
+        public LowCardinalityDictionaryWork(Page page)
+        {
+            this.page = requireNonNull(page, "page is null");
+            dictionarySizes = new int[channels.length];
+            blocks = new DictionaryBlock[channels.length];
+            int maxCardinality = 1;
+            for (int i = 0; i < channels.length; i++) {
+                Block block = page.getBlock(channels[i]);
+                verify(block instanceof DictionaryBlock, "Only dictionary blocks are supported");
+                blocks[i] = (DictionaryBlock) block;
+                int blockPositionCount = blocks[i].getDictionary().getPositionCount();
+                dictionarySizes[i] = blockPositionCount;
+                maxCardinality *= blockPositionCount;
+            }
+            this.maxCardinality = maxCardinality;
+        }
+
+        /**
+         * Returns combinations of all dictionaries ids for every position and populates
+         * samplePositions array with a single occurrence of every used combination
+         */
+        protected short[] calculatePositions(int positionCount, int[] samplePositions)
+        {
+            short[] combinationIds = new short[positionCount];
+
+            if (channels.length == 1) {
+                for (int i = 0; i < positionCount; i++) {
+                    combinationIds[i] = (short) blocks[0].getId(i);
+                    samplePositions[combinationIds[i]] = i;
+                }
+            }
+            else {
+                for (int i = 0; i < positionCount; i++) {
+                    combinationIds[i] = (short) blocks[0].getId(i);
+                }
+                for (int j = 1; j < channels.length; j++) {
+                    for (int i = 0; i < positionCount; i++) {
+                        combinationIds[i] *= dictionarySizes[j];
+                        combinationIds[i] += blocks[j].getId(i);
+                        samplePositions[combinationIds[i]] = i;
```

Comment:
Why not to do this only after all channels are processed? You would save a lot of assignments